arch-riscv: Improve widening/narrowing vectors overlap check

tommarin opened the pull request:
This PR improves the vector register groups overlap check in widening/narrowing 
instructions.

- Fix wrong illegal overlap condition between VS2 and VD vector register 
groups.
- Also check VS1 vector register group for overlap with VD in vector-vector 
instructions.
- Parametrize widening/narrowing factors in overlap check function to potentially 
handle more cases.

Fixes issue #442.

tommarin committed:
arch-riscv: Fix widening instructions vectors overlap check

This commit fixes the overlap check between VS2 and VD register groups
in vector widening instructions. While the narrowing instructions check
is correct, the widening one has to differentiate between two cases
(Vs2 EEW = 2*SEW and Vs2 EEW = SEW). In the first case, overlap is
allowed, as the EEW is the same as Vd. In the second case, the overlap
legality check has to be adapted to use the Vs2 EMUL to calculate the
boundaries. The rule has been derived again from Section 5.2 of RISC-V
"V" Vector Extension specifications, version 1.0.

The patch also includes some small code refactoring, e.g. using
already defined vlmul and constants for vector operands.

Fixes issue #442.

Change-Id: Ic87095fb9079e6c8f53b9a0d79fbf531a85dc71d

tommarin commented:
Considering the alignment to LMUL, the condition check in the case SEW->2\*SEW could be greatly simplified: the only illegal case is Vs2 == Vd. Still, it probably makes sense to keep it generic like in the original idea, to support for potential future cases (e.g. SEW->4\*SEW, as mentioned here: https://riscv.org/wp-content/uploads/2019/06/17.40-Vector_RISCV-20190611-Vectors.pdf). But in that case the calculation of num_microops will have to be made generic as well.

tommarin committed:
arch-riscv: Remove duplicate line

Change-Id: I32200aad5a59c9fd85f6ed783a4cebb841bf6ff1

hnpl commented on src/arch/riscv/isa/formats/vector_arith.isa:
```diff
@@ -79,38 +79,45 @@ let {{
-            const uint32_t num_microops = 1 << std::max<int64_t>(0, vtype_vlmul(machInst.vtype8) + 1);
+    def wideningOpRegisterConstraintChecks(code, src2_dw):
+        constraint_checks = '''
+            const uint32_t num_microops = 1 << std::max<int64_t>(0, vlmul + 1);
```
Is this specific to the case where `SEW` is widened to `2*SEW`?

tommarin replied:
Yes, indeed. As of now, no instruction widens to more than 2*SEW, but let me try to generalize it.

hnpl reviewed (commented):
Thanks for the contribution. I believe the issue in #442 would be fixed by this; however, I still have a couple of concerns.

- If I understand correctly, the `constraint_checks` code checks if the end of VD group overlaps with the start of VS group and if the end of VS2 group overlaps with the start of VD group. This totally makes sense to me. However, I'm not sure if `VD + num_microops - (1 << vlmul))` indicates the end of the VD group. Does this refer to the ["highest-numbered part" in the spec](https://github.com/riscv/riscv-v-spec/blob/2f68ef7256d6ec53e4d2bd7cb12862f406d64e34/v-spec.adoc?plain=1#L997)?
- How do you test the widening instructions with overlapping vector registers?

tommarin commented:
> Thanks for the contribution. I believe the issue in #442 would be fixed by this; however, I still have a couple of concerns.
> 
>     * If I understand correctly, the `constraint_checks` code checks if the end of VD group overlaps with the start of VS group and if the end of VS2 group overlaps with the start of VD group. This totally makes sense to me. However, I'm not sure if `VD + num_microops - (1 << vlmul))` indicates the end of the VD group. Does this refer to the ["highest-numbered part" in the spec](https://github.com/riscv/riscv-v-spec/blob/2f68ef7256d6ec53e4d2bd7cb12862f406d64e34/v-spec.adoc?plain=1#L997)?
> 
>     * How do you test the widening instructions with overlapping vector registers?

Thank you for reviewing the PR!
* Yes, that is the logic behind that boundary. We have to leave a space for overlap in the highest part, which depends on the size of the source vector group.
* I have checked that the commit solves the issue in two distinct cases, one being the example mentioned in the issue #442. I have also checked a custom-made example where the overlap happens (vd == vs2): this is properly detected.

I will try to generalize this function a little bit, e.g. de-hardcoding the widening factors.

hnpl commented on src/arch/riscv/isa/formats/vector_arith.isa:
```diff
@@ -79,38 +79,45 @@ let {{
+        if not src2_dw:
+            constraint_checks += '''
+            if (((vlmul < 0) && (VS2 == VD)) ||
+                ((vlmul >= 0) && (VS2 < VD + num_microops - (1 << vlmul)) &&
```
Please correct me if I'm wrong. I think it should be an `or` rather than an `and` for the boundary check, i.e.,

```python
((vlmul >= 0) && 
    ((VS2 < VD + num_microops - (1 << vlmul)) || (VD < VS2 + (1 << vlmul))))
```

This is because even if one of the start and the end of the VD group "overly" overlaps with VS2, it violates the boundary constraint.

tommarin replied:
To derive that condition, I started from the opposite one, which is where there is no (illegal) overlap.
- Any source vector higher than the top of the VD group (minus some overlap, sized as the VS2 group) is acceptable.
- Any source vector minor than the bottom of the VD group (minus the VS2 vector group size, to ensure no overlap at all) is acceptable.

If we negate this `or` conditions, we get an `and`: anything in the middle should be discarded.

A counterexample is, with LMUL=4, `vwmul.vv  v8, v0, v4`. `(VS2 < VD + 8 - 4)` is `true`, but this case is completely legal, as `(VD < VS2 + 4)` is `false`.

I tried to draw some of the cases, one real (LMUL=4, SEW->2\*SEW) and one hypothetical (LMUL=2, SEW->4\*SEW).
The red/orange area is the one where VS2 cannot sit because overlap is illegal. Orange area is also impossible because of alignment constraints.
In yellow I represented the area where both conditions are true in the original implementation, with boxes representing the register groups. In purple, the behavior of the new one (correct).

Please tell me if it makes sense. I will push a new commit with the generalization when ready.

![image](https://github.com/gem5/gem5/assets/55238129/12455243-bf71-4ab0-b56f-b6755304dd7f)

hnpl replied:
Thanks for the detailed explanation! I believe the condition you have is right.

I arrived at the same conclusion as yours when I think of this problem as the finding overlapping 1D segments problem. Basically we want VS2 to be either entirely on the left of VD (VS2.end < VD.start), or mostly to the right of VD (VS2.end > VD.start - num_microops - LMUL). The negation of this condition is exactly the same as yours.

saul44203 replied:
I believe this condition is right as well.  Apart from this, shouldn't we also check the vs1 vector register group for the vector-vector instructions (i.e. `.vv` and `.wv`)?

tommarin replied:
You're right. I've added another commit to fix that too.

tommarin committed:
arch-riscv: Generalize widening/narrowing vectors overlap check

As of now, the widening/narrowing vector register groups overlap check
always assumes a SEW multiplication factor equal to 2 (for either VD or
VS2). This commits aims at making this check more generic.

Change-Id: I4311fc3624cd324ccfdf2a1920a19efc85357120

tommarin committed:
arch-riscv: Check VS1 group for overlap when widening/narrowing

Currently, only the VS2 register group is checked for overlap with VD
when executing a widening/narrowing instruction. This commits extends
the check to VS1, when applicable (i.e. vector-vector operations).

Change-Id: I892b7717c01e25546fb41e05afbd08fc40c60c59
